### PR TITLE
`<functional>`: Fixes around `not_fn`

### LINKED ISSUE #4048: `<functional>`: `std::not_fn` can accept non-movable function
It seems that `std::not_fn` should reject the following code according to [[func.not.fn]](https://eel.is/c++draft/func.not.fn#2):

> _Mandates_: `is_constructible_v<FD, F> && is_move_constructible_v<FD>` is `true`.

[godbolt](https://godbolt.org/z/45sq67q55):

```cpp
#include <functional>

struct OnlyCopyableFun {
  OnlyCopyableFun() = default;
  OnlyCopyableFun(const OnlyCopyableFun&) = default;
  OnlyCopyableFun(OnlyCopyableFun&&) = delete;
  bool operator()(auto) const;
};

int main() {
  OnlyCopyableFun f;
  auto nf = std::not_fn(f); // only ill-formed in libc++
}
```
Feel free to correct me if I've missed something.
Not quite sure why these standard call wrapper factories (`std::bind_front`, `std::bind`, etc.) _require_ all argument types to be move-constructible, although the call wrapper produced in the above example is _still_ move-constructible as its underlying type has a copy constructor. 
It seems to me that just requiring `is_constructible_v<FD, F>` is enough, which is what the range adaptor object does.

### PR DESCRIPTION
**Pull request by frederick-vs-ja**
Make the return type (`_Not_fn`) perfect forwarding call wrapper since C++20, which is a change in WG21-P0356R5. 

Currently this PR leaves `_Not_fn` as-is in C++17 mode, because the return type was nearly fully specified in C++17 ([N4659 [func.not_fn]/1](https://timsong-cpp.github.io/cppwp/n4659/func.not_fn#1)), which means that
1. due to the absence of deleted `operator()` overloads, an invocation to a non-const value of the return type might select a const overload, and
2. due to the use of `decltype(!declval<invoke_result_t<...>>())`, if the `operator!` takes a non-movable return type of `std::invoke(args...)` by value, the return type of `operator()` would be ill-formed, while `!std::invoke(args...)` being well-formed.

In C++20 the first case should be ill-formed (in an SFINAE-friendly way), and the second case should be well-formed.

Also make `not_fn` reject non-move-constructible types. Fixes #4048.

I don't know why call wrappers (except for `std::function`) are still required to be _Cpp17MoveConstructible_ in C++17 and later. It seems that the restrictions can be lifted.

### CODE REVIEW

**Comment by hewillk** [comment]
You should probably also add a `static_assert` for `is_constructible_v<FD, F>` to avoid hard errors in the function body:

```cpp
struct OnlyMovableFun {
  OnlyMovableFun() = default;
  OnlyMovableFun(const OnlyMovableFun&) = delete;
  OnlyMovableFun(OnlyMovableFun&&) = default;
  bool operator()(auto) const;
};

int main() {
  OnlyMovableFun f;
  auto nf = std::not_fn(f); // should be static_assert?
}
```

**Comment by CaseyCarter** [comment]
> You should probably also add a `static_assert` for `is_constructible_v<FD, F>` to avoid hard errors in the function body:

That wouldn't "avoid hard errors", it would transform a hard error into a different hard error.

File: tests/std/tests/Dev11_0535636_functional_overhaul/test.cpp [hunk 1]
```diff
@@ -2356,6 +2356,62 @@ _CONSTEXPR20 bool test_not_fn() {
...
+        return {};
+    }
+};
+
+using NegatedConstOnlyFunctor       = decltype(not_fn(ConstOnlyFunctor{}));
+using NegatedConstOnlyBooleanTester = decltype(not_fn(ConstOnlyBooleanTester{}));
+using NegatedGetPinnedNegatable     = decltype(not_fn(GetPinnedNegatable{}));
+
+#if _HAS_CXX20
+constexpr bool not_fn_is_perfert_forwarding = true;
```
Comment (StephanTLavavej) [inline tests/std/tests/Dev11_0535636_functional_overhaul/test.cpp:2397]: "perfert" typo.

File: tests/std/tests/Dev11_0535636_functional_overhaul/test.cpp [hunk 2]
```diff
@@ -2356,6 +2356,62 @@ _CONSTEXPR20 bool test_not_fn() {
...
+#else // ^^^ _HAS_CXX20 / !_HAS_CXX20 vvv
+constexpr bool not_fn_is_perfert_forwarding = false;
+#endif // ^^^ !_HAS_CXX20 ^^
```
Comment (StephanTLavavej) [inline tests/std/tests/Dev11_0535636_functional_overhaul/test.cpp:2400]: Missing one final arrow.

**Comment by StephanTLavavej** [comment]
Thanks - this compile-time change is low-risk enough that I think we need only one maintainer approval.

**Comment by StephanTLavavej** [comment]
I'm mirroring this to the MSVC-internal repo - please notify me if any further changes are pushed.

**Comment by StephanTLavavej** [comment]
Thanks for noticing and fixing this issue in `not_fn`! :hammer_and_wrench: :tada: :heart_eyes_cat:

**Pull request merged**